### PR TITLE
ENH: asymptotic expansion for multivariate t entropy

### PR DESCRIPTION
**Pull request by OmarManzoor**
#### Reference issue

Towards: gh-18093

#### What does this implement/fix?

- Adds asymptotic expansion for the entropy of multivariate t distribution when considering large values of df.
- Adds some tests

#### Additional information

CC: @dschmitz89

### CODE REVIEW

**Comment by OmarManzoor** [comment]
image

<Details>

<Summary>Code to generate relative error</Summary>

```python
class MultivariateT_Entropy:
    def _process_parameters(self, loc, shape, df):
        if loc is None and shape is None:
            loc = np.asarray(0, dtype=float)
            shape = np.asarray(1, dtype=float)
            dim = 1
        elif loc is None:
            shape = np.asarray(shape, dtype=float)
            if shape.ndim < 2:
                dim = 1
            else:
                dim = shape.shape[0]
            loc = np.zeros(dim)
        elif shape is None:
            loc = np.asarray(loc, dtype=float)
            dim = loc.size
            shape = np.eye(dim)
        else:
            shape = np.asarray(shape, dtype=float)
            loc = np.asarray(loc, dtype=float)
            dim = loc.size

        if dim == 1:
            loc = loc.reshape(1)
            shape = shape.reshape(1, 1)

        if loc.ndim != 1 or loc.shape[0] != dim:
            raise ValueError("Array 'loc' must be a vector of length %d." %
                              dim)
        if shape.ndim == 0:
            shape = shape * np.eye(dim)
        elif shape.ndim == 1:
            shape = np.diag(shape)
        elif shape.ndim == 2 and shape.shape != (dim, dim):
            rows, cols = shape.shape
            if rows != cols:
                msg = ("Array 'cov' must be square if it is two dimensional,"
                        " but cov.shape = %s." % str(shape.shape))
            else:
                msg = ("Dimension mismatch: array 'cov' is of shape %s,"
                        " but 'loc' is a vector of length %d.")
                msg = msg % (str(shape.shape), len(loc))
            raise ValueError(msg)
        elif shape.ndim > 2:
            raise ValueError("Array 'cov' must be at most two-dimensional,"
                              " but cov.ndim = %d" % shape.ndim)

        # Process degrees of freedom.
        if df is None:
            df = 1
        elif df <= 0:
            raise ValueError("'df' must be greater than zero.")
        elif np.isnan(df):
            raise ValueError("'df' is 'nan' but must be greater than zero or 'np.inf'.")

        return dim, loc, shape, df

    def _entropy(self, dim, df=1, shape=1):
        if df == np.inf:
            return multivariate_normal(None, cov=shape).entropy()
        
        shape_info = _PSD(shape)
        halfsum = 0.5 * (dim + df)
        half_df = 0.5 * df
        return (-gammaln(halfsum) + gammaln(half_df)
                + 0.5 * dim * np.log(df * np.pi) + halfsum
                * (psi(halfsum) - psi(half_df))
                + 0.5 * shape_info.log_pdet)
    
    def entropy(self, loc=None, shape=1, df=1):
        dim, loc, shape, df = self._process_parameters(None, shape, df)
        return self._entropy(dim, df, shape)

    def asymptotic(self, loc=None, shape=1, df=1):
        dim, loc, shape, df = self._process_parameters(None, shape, df)
        shape_info = _PSD(shape)
        shape_term = 0.5 * shape_info.log_pdet
        return (
            1.41894 * dim + dim / df
            - dim * (dim - 2) * df**-2.0 / 4
            + dim**2 * (dim - 2) * df**-3.0 / 6
            + dim * (-3 * dim**3 + 8 * dim**2 - 8) * df**-4.0 / 24
            + shape_term
        )

@np.vectorize
def mul_t_mpmath_entropy(dim, df=1, shape=1):
    dim = mp.mpf(dim)
    df = mp.mpf(df)
    shape = mp.mpf(shape)
    halfsum = (dim + df) / 2
    half_df = df / 2
    
    return float(
        -mp.loggamma(halfsum) + mp.loggamma(half_df)
        + 0.5 * dim * mp.log(df * mp.pi)
        + halfsum * (mp.digamma(halfsum) - mp.digamma(half_df))
        + 0.0
    )

x = np.logspace(0, 15, 1000)
dim = 1.0
mul_t = MultivariateT_Entropy()

reference = np.array(mul_t_mpmath_entropy(dim, x), np.float64)
regular = np.array([mul_t.entropy(shape=1.0, df=_df) for _df in x], np.float64)
asymptotic_res = np.array([mul_t.asymptotic(shape=1.0, df=_df) for _df in x], np.float64)

plt.loglog(x, np.abs((regular - reference)/reference), label="regular", ls='dashed')
plt.loglog(x, np.abs((asymptotic_res - reference)/reference), label="asymptotic", ls='dotted')
plt.axvline(5e10, 0, 1, c='k', label="Threshold")
plt.title("Relative error: multivariate t entropy")
plt.legend(loc='upper left')
plt.xlabel("$a$")
plt.ylabel("$h$")
plt.show()
```
</Details>

**Comment by mdhaber** [comment]
I don't plan to review this in detail, but to prepare this for another reviewer, please try to explain or correct the remaining ~1e-5 error in the asymptotic version. If it is a correct asymptotic expansion, we would expect the error up go toward machine precision as `df` increases.

**Review by dschmitz89 — COMMENTED** [review]
This looks very good already. :)

 Please rerun the accuracy check for finding the threshold for different dimensions, so we can assess if 400 is a good cutoff in general still. Currently this is not testing the multivariate but univariate case.

**Comment by OmarManzoor** [comment]
@dschmitz89 

### Dim = 3
image

<Details>

```python
x = np.logspace(0, 15, 1000)
dim = 5.0
shape = np.eye(5, 5)
mul_t = MultivariateT_Entropy()

reference = np.array(mul_t_mpmath_entropy(dim, x), np.float64)
regular = np.array([mul_t.entropy(shape=shape, df=_df) for _df in x], np.float64)
asymptotic_res = np.array([mul_t.asymptotic(shape=shape, df=_df) for _df in x], np.float64)

plt.loglog(x, np.abs((regular - reference)/reference), label="regular", ls='dashed')
plt.loglog(x, np.abs((asymptotic_res - reference)/reference), label="asymptotic", ls='dotted')
plt.axvline(400, 0, 1, c='k', label="Threshold")
plt.title("Relative error: multivariate t entropy")
plt.legend(loc='upper left')
plt.xlabel("$a$")
plt.ylabel("$h$")
plt.show()
```

</Details>

### Dim = 5
image

<Details>

```python
x = np.logspace(0, 15, 1000)
dim = 5.0
shape = np.eye(5, 5)
mul_t = MultivariateT_Entropy()

reference = np.array(mul_t_mpmath_entropy(dim, x), np.float64)
regular = np.array([mul_t.entropy(shape=shape, df=_df) for _df in x], np.float64)
asymptotic_res = np.array([mul_t.asymptotic(shape=shape, df=_df) for _df in x], np.float64)

plt.loglog(x, np.abs((regular - reference)/reference), label="regular", ls='dashed')
plt.loglog(x, np.abs((asymptotic_res - reference)/reference), label="asymptotic", ls='dotted')
plt.axvline(400, 0, 1, c='k', label="Threshold")
plt.title("Relative error: multivariate t entropy")
plt.legend(loc='upper left')
plt.xlabel("$a$")
plt.ylabel("$h$")
plt.show()
```

</Details>

### Dim = 10

image


### Dim = 100

image

**Comment by OmarManzoor** [comment]
I think if we set the limit to this 
`dim*100*(4/(np.log10(dim) + 1))`
the results look much better

**Comment by dschmitz89** [comment]
Yes, that looks better. Please update the threshold and tests accordingly :)

File: scipy/stats/_multivariate.py
Comment (dschmitz89) [inline scipy/stats/_multivariate.py:4592]: I am not sure if we need `_lazywhere` here. Can `entropy` be called with several parameter sets for other multivariate distributions? I cannot test at the moment as I am only on my phone. If `entropy` only works for one parameter set, a simple if/else would be enough.

File: scipy/stats/_multivariate.py
Comment (mdhaber) [inline scipy/stats/_multivariate.py:4589]: This is currently returning a 0d array. Ensure that it returns a scalar:
```suggestion
            )[()]
```

File: scipy/stats/_multivariate.py
Comment (mdhaber) [inline scipy/stats/_multivariate.py:4582]: To double-check, I copy-pasted the regular formula above and made appropriate replacements. I got something equivalent.
```
asymptotic expansion -gammaln((m+d)/2) + gammaln(d/2) + m/2*log(d*pi) + (m+d)/2 * (digamma((m+d)/2) - digamma(d/2))
```
Wolfram Alpha gives
image

before some higher order terms, and that seems to be what you've coded below. I'm not concerned about `dim` to high powers because without sparse arrays it's impossible for `dim` to be high enough to cause overflows.

File: scipy/stats/_multivariate.py
Comment (mdhaber) [inline scipy/stats/_multivariate.py:4591]: It looks like you've found a good heuristic, but I'd be more ready to merge this if you were to instead add the fifth-order terms to the asymptotic expansion. That would probably create overlap between the regions where each expression is accurate, so you wouldn't need a heuristic.

image

File: scipy/stats/tests/test_multivariate.py
Comment (mdhaber) [inline scipy/stats/tests/test_multivariate.py:2629]: I confirmed these with the following code:

<details>

```python3
from mpmath import mp
data = [
    (10, 1, 1.5212624929756808, 1e-15),
    (100, 1, 1.4289633653182439, 1e-13),
    (500, 1, 1.420939531869349, 1e-14),
    (1e20, 1, 1.4189385332046727, 1e-15),
    (1e100, 1, 1.4189385332046727, 1e-15),
    (10, 10, 15.069150450832911, 1e-15),
    (1000, 10, 14.19936546446673, 1e-13),
    (1e20, 10, 14.189385332046728, 1e-15),
    (1e100, 10, 14.189385332046728, 1e-15),
    (10, 100, 148.28902883192654, 1e-15),
    (1000, 100, 141.99155538003762, 1e-14),
    (1e20, 100, 141.8938533204673, 1e-15),
    (1e100, 100, 141.8938533204673, 1e-15),
]

dfs, dims, refs, _ = np.array(data).T

def reference(dim, df):
    dim = mp.mpf(dim)
    df = mp.mpf(df)
    halfsum = (dim + df) / 2
    half_df = df / 2

    def gammaln(x):
        return mp.log(mp.gamma(x))

    psi = mp.digamma
    return float(-gammaln(halfsum) + gammaln(half_df) + dim / 2 * mp.log(df * mp.pi) + halfsum * (psi(halfsum) - psi(half_df)))

for df, dim, ref in zip(dfs, dims, refs):
    res = reference(dim, df)
    print(res - ref)
```

</details>

File: scipy/stats/tests/test_multivariate.py
Comment (mdhaber) [inline scipy/stats/tests/test_multivariate.py:2627]: As a matter of principle, please prefer division by integers rather than decimal expressions when working with mpmath. It doesn't matter here, but it's a good habit.

```suggestion
        #     halfsum = (dim + df)/2
        #     half_df = df/2
        #
        #     return float(
        #         -mp.loggamma(halfsum) + mp.loggamma(half_df)
        #         + dim / 2 * mp.log(df * mp.pi)
        #         + halfsum * (mp.digamma(halfsum) - mp.digamma(half_df))
```

File: scipy/stats/tests/test_multivariate.py
Comment (mdhaber) [inline scipy/stats/tests/test_multivariate.py:2628]: Please add one test that confirms that the shape term is implemented correctly. That is, add a test with nontrivial covariance matrix that substantially affects the value of the entropy.

For instance, a simpler (but less accurate) approach to this PR would have been to use the fact that the `multivariate_t` entropy approaches the `multivariate_normal` entropy as the number of degrees of freedom gets bigger. Your test could simply compare the two entropies with a nontrivial covariance matrix and large degrees of freedom.

Then, you could confirm that the implementation of the shape term in the non-asymptotic formula is correct by comparing the `multivariate_t.entropy` with slightly different `df` - on either side of the threshold that switches between them. They should be nearly the same.

**Review by mdhaber — CHANGES_REQUESTED** [review]
Just a few things here, then we can merge.

File: scipy/stats/_multivariate.py
Comment (OmarManzoor) [reply scipy/stats/_multivariate.py:4591]: Even if I add the fifth order term we would still need a threshold right?

Comment (mdhaber) [reply scipy/stats/_multivariate.py:4591]: Yes. I suppose we could call a threshold a heuristic, too, but it's pretty standard practice. I think we have use a fixed threshold in all of the other asymptotic expansion PRs.

Comment (OmarManzoor) [reply scipy/stats/_multivariate.py:4591]: Right, so that means using a fifth order term would enable us to specify a fixed threshold?

Comment (OmarManzoor) [reply scipy/stats/_multivariate.py:4591]: Thanks!

Comment (mdhaber) [reply scipy/stats/_multivariate.py]: I hoped that adding the extra term would extend the accurate portion of the asymptotic approximation a lot further to the left. Oh well. 

> Should we only consider dim to be up till a specific value and ignore higher ones?

I think that is safe because currently `multivariate_t` accepts only dense arrays, so the memory required scales with `dim**2` right? Since no bug reports have even been filed about `multivariate_t` entropy precision, I am not sure whether anyone will make use of this extra precision, let alone users with a petabyte of RAM. If that's right, I think it's OK to stop looking beyond `dim > 1e7`.

Comment (OmarManzoor) [reply scipy/stats/_multivariate.py]: Sounds good.

**Review by mdhaber — APPROVED** [review]
OK, if you would post the latest plots with the new heuristic, I think we can merge this.

File: scipy/stats/_multivariate.py
Comment (mdhaber) [inline scipy/stats/_multivariate.py:4592]: I'll commit this after tests finish with `[skip ci]`.

```suggestion
        # observed to preserve accuracy up to at least `dim = 1e7`. See gh-18465.
        threshold = 200 * dim
```

**Review by mdhaber — APPROVED** [review]

**Comment by OmarManzoor** [comment]
@mdhaber 

image

Code
<Details>

```python
class MultivariateT_Entropy:
    def _process_parameters(self, loc, shape, df):
        if loc is None and shape is None:
            loc = np.asarray(0, dtype=float)
            shape = np.asarray(1, dtype=float)
            dim = 1
        elif loc is None:
            shape = np.asarray(shape, dtype=float)
            if shape.ndim < 2:
                dim = 1
            else:
                dim = shape.shape[0]
            loc = np.zeros(dim)
        elif shape is None:
            loc = np.asarray(loc, dtype=float)
            dim = loc.size
            shape = np.eye(dim)
        else:
            shape = np.asarray(shape, dtype=float)
            loc = np.asarray(loc, dtype=float)
            dim = loc.size

        if dim == 1:
            loc = loc.reshape(1)
            shape = shape.reshape(1, 1)

        if loc.ndim != 1 or loc.shape[0] != dim:
            raise ValueError("Array 'loc' must be a vector of length %d." %
                              dim)
        if shape.ndim == 0:
            shape = shape * np.eye(dim)
        elif shape.ndim == 1:
            shape = np.diag(shape)
        elif shape.ndim == 2 and shape.shape != (dim, dim):
            rows, cols = shape.shape
            if rows != cols:
                msg = ("Array 'cov' must be square if it is two dimensional,"
                        " but cov.shape = %s." % str(shape.shape))
            else:
                msg = ("Dimension mismatch: array 'cov' is of shape %s,"
                        " but 'loc' is a vector of length %d.")
                msg = msg % (str(shape.shape), len(loc))
            raise ValueError(msg)
        elif shape.ndim > 2:
            raise ValueError("Array 'cov' must be at most two-dimensional,"
                              " but cov.ndim = %d" % shape.ndim)

        # Process degrees of freedom.
        if df is None:
            df = 1
        elif df <= 0:
            raise ValueError("'df' must be greater than zero.")
        elif np.isnan(df):
            raise ValueError("'df' is 'nan' but must be greater than zero or 'np.inf'.")

        return dim, loc, shape, df

    def _entropy(self, dim, df=1, shape=1):
        if df == np.inf:
            return multivariate_normal(None, cov=shape).entropy()
        
        #shape_info = _PSD(shape)
        halfsum = 0.5 * (dim + df)
        half_df = 0.5 * df
        return (-gammaln(halfsum) + gammaln(half_df)
                + 0.5 * dim * np.log(df * np.pi) + halfsum
                * (psi(halfsum) - psi(half_df))
                + 0.0 )#0.5 * shape_info.log_pdet)
    
    def entropy(self, loc=None, shape=1, df=1):
        dim, loc, shape, df = self._process_parameters(None, shape, df)
        return self._entropy(dim, df, shape)

    def asymptotic(self, loc=None, shape=1, df=1):
        dim, loc, shape, df = self._process_parameters(None, shape, df)
        #shape_info = _PSD(shape)
        #shape_term = 0.5 * shape_info.log_pdet
        return (
            dim * stats.norm._entropy() + dim / df
            - dim * (dim - 2) * df**-2.0 / 4
            + dim**2 * (dim - 2) * df**-3.0 / 6
            + dim * (-3 * dim**3 + 8 * dim**2 - 8) * df**-4.0 / 24
            + dim**2 * (3 * dim**3 - 10 * dim**2 + 16) * df**-5.0 / 30
            + 0.0
        )

@np.vectorize
def mul_t_mpmath_entropy(dim, df=1):
    dim = mp.mpf(dim)
    df = mp.mpf(df)
    halfsum = (dim + df) / 2
    half_df = df / 2
    
    return float(
        -mp.loggamma(halfsum) + mp.loggamma(half_df)
        + dim / 2 * mp.log(df * mp.pi)
        + halfsum * (mp.digamma(halfsum) - mp.digamma(half_df))
        + 0.0
    )

# Plotting code
x = np.logspace(0, 15, 1000)
dims = [1.0, 10.0, 50.0, 100.0, 500.0, 1000.0, 5000.0, 10000.0, 100000.0]
figure, axis = plt.subplots(3, 3, figsize=(10, 9))
figure.tight_layout(pad=3.0)
k = 0
for i in range(3):
    for j in range(3):
        dim = dims[k]
        d = int(dim)
        shape = np.eye(d, d)
        mul_t = MultivariateT_Entropy()
        reference = np.array(mul_t_mpmath_entropy(dim, x), np.float64)
        regular = np.array([mul_t.entropy(shape=shape, df=_df) for _df in x], np.float64)
        asymptotic_res = np.array([mul_t.asymptotic(shape=shape, df=_df) for _df in x], np.float64)

        ax = axis[i, j]
        ax.loglog(x, np.abs((regular - reference)/reference), label="regular", ls='dashed')
        ax.loglog(x, np.abs((asymptotic_res - reference)/reference), label="asymptotic", ls='dotted')
        ax.axvline(d * 200, 0, 1, c='k', label="Threshold")
        ax.set_title(f"Relative error for Dim={d}", fontsize=8)
        ax.set_xlabel("$a$", fontsize=8)
        ax.set_ylabel("$h$", fontsize=8)
        k += 1

plt.show()
```

</Details>

**Comment by OmarManzoor** [comment]
> @mdhaber
> 
> image
> 
> Code

Aren't we going a bit to the right with this?

**Comment by mdhaber** [comment]
Yes, it looks like the old one tracked better. I'll let you decide what to do.

**Comment by dschmitz89** [comment]
My bad. Indeed, let's use the hold threshold.

**Comment by OmarManzoor** [comment]
@mdhaber @dschmitz89 
I adjusted the threshold slightly using **_np.log_** instead of **_np.log10_**

image

<Details>

```python
# Plotting
x = np.logspace(0, 15, 1000)
dims = [1.0, 10.0, 50.0, 100.0, 500.0, 1000.0, 5000.0, 10000.0, 100000.0]
figure, axis = plt.subplots(3, 3, figsize=(10, 9))
figure.tight_layout(pad=3.0)
k = 0
for i in range(3):
    for j in range(3):
        dim = dims[k]
        d = int(dim)
        shape = np.eye(d, d)
        mul_t = MultivariateT_Entropy()
        reference = np.array(mul_t_mpmath_entropy(dim, x), np.float64)
        regular = np.array([mul_t.entropy(shape=shape, df=_df) for _df in x], np.float64)
        asymptotic_res = np.array([mul_t.asymptotic(shape=shape, df=_df) for _df in x], np.float64)

        ax = axis[i, j]
        v = 4/(np.log(d) + 1)
        ax.loglog(x, np.abs((regular - reference)/reference), label="regular", ls='dashed')
        ax.loglog(x, np.abs((asymptotic_res - reference)/reference), label="asymptotic", ls='dotted')
        ax.axvline(d*100*v, 0, 1, c='k', label="Threshold")
        ax.set_title(f"Relative error for Dim={d}", fontsize=8)
        ax.set_xlabel("$a$", fontsize=8)
        ax.set_ylabel("$h$", fontsize=8)
        k += 1

plt.show()
```

</Details>

File: scipy/stats/_multivariate.py
Comment (mdhaber) [inline scipy/stats/_multivariate.py:4592]: Again, no need to re-run CI, but I think this deserves a comment. I will commit with `[skip ci]` after tests pass.
```suggestion
        # preserves ~12 digits accuracy up to at least `dim=1e5`. See gh-18465.
        threshold = dim * 100 * 4 / (np.log(dim) + 1)
```

**Pull request merged**
